packages/protocol/docs/how_taiko_proves_blocks.md

```diff
@@ -0,0 +1,360 @@
+
+- `blockhash`: EVM allows access to the most recent 256 block hashes. All these hashes are available inside the plonk lookup table. ZKP must prove that the lookup table is consistent with L2's historical values.
+  - `blockhash(block.number - 1)`, has the same value as in the block header and is also the same value as the parent block's hash on L2.
+  - The other 255 hashes, `blockhash(block.number - 256)` to `blockhash(block.number - 2)` are checked in the anchor transaction to simplify circuits. Therefore, as long as the anchor transaction is zk-proven, these 255 ancestor hashes are proven indirectly.
+- `block.basefee`: verified to be the correct value in the anchor transaction.
+- `block.chainid`: this field is also checked by the anchor transaction, so no extra ZKP circuits are required.
+- `block.coinbase`: ZKP must verify the value must be the same as `meta.beneficiary`. Again, the metadata hash is part of the ZK instance.
+- `block.difficulty`: this is now the same as `block.prevrandao`, so we only check `block.prevrandao`.
+- `block.gaslimit`: ZKP must verify this value must equal `meta.gasLimit`.
+- `block.number`: this must be checked against the block header and `meta.id`.
+- `block.prevrandao`: this must be checked against the `mixHash` field in the L2 block header and `meta.mixHash`.
+- `block.timestamp`: this must be checked against the `timestamp` field in the L2 block header and `meta.proposedAt`.
+
+### Block Header
+
+Not all block header data is available in the L1 contracts; therefore, the ZKP must verify the integrity of the block header and ensure the block header hashes into the same value as `evidence.blockHash`, which is part of the ZK instance input.
+
+```solidity
+struct BlockHeader {
+  bytes32 parentHash;
+  bytes32 ommersHash;
+  address beneficiary;
+  bytes32 stateRoot;
+  bytes32 transactionsRoot;
+  bytes32 receiptsRoot;
+  bytes32[8] logsBloom;
+  uint256 difficulty;
+  uint128 height;
+  uint64 gasLimit;
+  uint64 gasUsed;
+  uint64 timestamp;
+  bytes extraData;
+  bytes32 mixHash;
+  uint64 nonce;
+  uint256 baseFeePerGas;
+  bytes32 withdrawalsRoot;
+}
+```
+
+In addition, ZKP must also prove the following:
+
+- `parentHash` must be the same as `evidence.parentHash`.
+- `ommersHash` must be the keccak256 of `[]`,or `0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347`.
+- `beneficiary` must be `meta.beneficiary` (duplicated, as stated above).
+- `logsBloom` must be a `bytes32[8]` with all zeros.
+- `difficulty` == 0.
+- `height` must be `meta.id`.
+- `gasLimit` == `meta.gasLimit`.
+- `gasUsed` must be the sum of all gas used by all valid transactions in the block.
+- `timestamp` == `meta.proposedAt`.
+- `extraData` == "".
+- `mixHash` == `meta.mixHash`.
+- `nonce` == 0.
+- `baseFeePerGas` == the calculated EIP-1559 style base fee.
+- `withdrawalsRoot` == The withdrawals MPT root corresponding to the deposits list.
```

Comment:
Oh we store that hash directly in our L2 block header? Maybe better to then just store the empty MPT root hash instead for now so we are at least compatible with standard Ethereum block headers?

Reply:
I think in current codebase, yes. Here is the calculation of the `depositRoot`: https://github.com/taikoxyz/taiko-mono/blob/main/packages/protocol/contracts/L1/libs/LibEthDepositing.sol#L107 which the corresponding `header.withdrawRoot` needs to equal to.

Reply:
Fair enough for now, but something I think we should change in the future.

Reply:
One reason we store the hash of a list of `EthDeposit` is to reduce cost -- hashing the MPT root requires more `keccek` ops.

Reply:
Oh yeah I'm not saying we should modify the solidity code, the `depositsRoot` of that data is just a circuit input so we can do whatever we want there, no problem. It's just that the value stored in the L2 block header for this should ideally be the MPT root of this data, and that MPT root would be calculated in the circuits (or we could just leave it be the empty MPT tree hash for now).

Reply:
so `depositsRoot` is still the flat hash of the _deposits_, and `withdrawalsRoot` (which is in the header) is the MPT root of the same deposit list?

Reply:
Not sure if its hard for circuits? but it will reduce lots of changes which we've made in go-ethereum.

Reply:
Should be very similar to creating the tx/log trie, so should be fine. I honestly also think an acceptable (perhaps intermediate) solution is just setting the withdrawals root to simply the empty MPT root in the block header, it's not like the value is that useful. And there are ways to make depositing ETH even cheaper than this mechanism, so maybe won't even be used in the future. Is having the actual data in the MPT trie root important for the node?

> but it will reduce lots of changes which we've made in go-ethereum.

Can you explain a bit more why it required a lot of changes?

Reply:
Brecht, it would be ideal to utilize the `withdrawalsRoot` as it is because the go-ethereum code already supports this feature, requiring no additional action on our part. When I say *use the withdrawalsRoot as is*, I mean:

- The structure of an _EthDeposit_ should be defined in the same way as an ETH2 withdrawal. (@davidtaikocha  Could you please provide the definition)?

- We should hash these structures in the same manner as in ETH2.

Currently, our _EthDeposit_ definition is independent of the ETH2 withdrawal data structure, and we hash all deposits as a simple array instead of a tree. That's why David mentioned that we had to modify the go-ethereum codebase with additional code. In other words, we only re-use the `withdrawalsRoot`, not the corresponding go-ethereum code.

Reply:
here is the definition:
<img width="1014" alt="image" src="https://github.com/taikoxyz/taiko-mono/assets/104078303/deed64ca-294d-40d7-aa90-63dc611ac967">

Reply:
> Brecht, it would be ideal to utilize the `withdrawalsRoot` as it is because the go-ethereum code already supports this feature, requiring no additional action on our part. When I say _use the withdrawalsRoot as is_, I mean:
> 
> * The structure of an _EthDeposit_ should be defined in the same way as an ETH2 withdrawal. (@davidtaikocha  Could you please provide the definition)?
> * We should hash these structures in the same manner as in ETH2.
> 
> Currently, our _EthDeposit_ definition is independent of the ETH2 withdrawal data structure, and we hash all deposits as a simple array instead of a tree. That's why David mentioned that we had to modify the go-ethereum codebase with additional code. In other words, we only re-use the `withdrawalsRoot` field in the block header, not the corresponding go-ethereum code.

and i think right now there is at least one different between the protocol's `EthDeposit` and eth2's `Withdrawal`, which is the unit of `EthDeposit.Amount` in protocol is `wei`, but when it comes to eth2's `Withdrawal.Amount`, the unit is `gwei`.

Reply:
My main question is still if `withdrawalsRoot` is actually used in the node to do anything? If you change it to the root of an empty tree would everything still work on the node side (except not being able to query what withdrawals happened in a block, which seems okay to me)?

Of course somewhere on the node side we'll still have to get the ETH deposits from the smart contract (this could be easy just using data from events or something). Then I would think it takes that data and translates it into whatever the data structure is already used in the node to actually execute these withdrawals so I would expect the actual node changes to be quite small. I assume here that that data structure is actually totally unrelated to the withdrawals MPT, which I would think is just be there to make the data accessible in an efficient way (like the tx tree), and is otherwise unused in the node while actually executing the block.

Reply:
> My main question is still if withdrawalsRoot is actually used in the node to do anything?
I think it is actually used by the client to perform minting Ether out of thin air. @davidtaikocha

Reply:
> > My main question is still if withdrawalsRoot is actually used in the node to do anything?
> 
> I think it is actually used by the client to perform minting Ether out of thin air. @davidtaikocha

right, its used to mint real L2 ETHs, if there are some `depositsProcessed` in the corresponding `blockMetaData`

Reply:
So the actual data is stored as a simple list in `depositsProcessed` in `blockMetaData`? So the node does not use the withdrawals MPT data directly then?

I'm just trying to figure out what it's actually used for (except making it efficient to check what withdrawals were done after the block has already been executed, so not important for the execution itself). The transactions in a block are also made available in the tx tree, but I don't think the node actually ever uses that tree (I think it's specifically made to just put in the block header).

Reply:
Yes, the actual data is stored in protocol, when mining a new block, the node need to create a ETH2 withdrawals list based on the corresponding `depositsProcessed` and process them, after that the node will use the created `ETH2 withdrawals list` to calculate a `block.withdrawalsRoot`.

Reply:
Okay thanks! So I would think node changes can be kept to a minimum then because we do indeed only have to do the `depositsProcessed` -> `ETH2 withdrawals list` translation somewhere and then everything else can remain the same for the node as far as I understand.

Then I would also think that we have a lot of freedom for the value we put in the block header for `withdrawalsRoot`.

Ideally of course it's the standard MPT root value, and I would say let's jet set it to that value for now to minimize differences.

Forcing it to the empty MPT root for each block seems also okay for me, the only thing we break by doing that is that it doesn't allow using Merkle proofs anymore to check which withdrawal was done in a block. So if we want to save the work building this MPT tree in the circuit that sounds fine to me. This also requires almost no node changes.